### PR TITLE
[FLINK-31397] Introduce write-once hash lookup store

### PR DESCRIPTION
**Pull request by JingsongLi**
Introduce interface for lookup changelog producer:
```
/**
 * A key-value store for lookup, key-value store should be single binary file written once and ready
 * to be used. This factory provide two interfaces:
 *
 * <ul>
 *   <li>Writer: written once to prepare binary file.
 *   <li>Reader: lookup value by key bytes.
 * </ul>
 */
public interface LookupStoreFactory {

    LookupStoreWriter createWriter(File file) throws IOException;

    LookupStoreReader createReader(File file) throws IOException;
}
```
We can convert remote columnar data to local lookup store, and ready to be used to lookup.

### CODE REVIEW

File: flink-table-store-common/src/main/java/org/apache/flink/table/store/lookup/LookupStoreReader.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+package org.apache.flink.table.store.lookup;
+
+import java.io.Closeable;
+import java.io.IOException;
+
+/** Reader, lookup value by key bytes. */
+public interface LookupStoreReader extends Closeable {
+
+    byte[] lookup(byte[] key) throws IOException;
```
Comment (SteNicholas) [inline flink-table-store-common/src/main/java/org/apache/flink/table/store/lookup/LookupStoreReader.java:27]: Adds the java doc for interface methods.

File: flink-table-store-common/src/main/java/org/apache/flink/table/store/lookup/LookupStoreWriter.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+package org.apache.flink.table.store.lookup;
+
+import java.io.Closeable;
+import java.io.IOException;
+
+/** Writer to prepare binary file. */
+public interface LookupStoreWriter extends Closeable {
+
+    void put(byte[] key, byte[] value) throws IOException;
```
Comment (SteNicholas) [inline flink-table-store-common/src/main/java/org/apache/flink/table/store/lookup/LookupStoreWriter.java:27]: Ditto.

File: flink-table-store-common/src/main/java/org/apache/flink/table/store/lookup/hash/HashLookupStoreWriter.java [hunk 3]
```diff
@@ -0,0 +1,486 @@
...
+        LOG.info("Number of keys: {}", keyCount);
+        LOG.info("Number of values: {}", valueCount);
+
+        // Prepare files to merge
+        List<File> filesToMerge = new ArrayList<>();
+
+        try {
+
+            // Write metadata file
+            File metadataFile = new File(tempFolder, "metadata.dat");
```
Comment (SteNicholas) [inline flink-table-store-common/src/main/java/org/apache/flink/table/store/lookup/hash/HashLookupStoreWriter.java:178]: Why do the process including writing metadata file, building index file, statsing collisions, adding data files, and merging and writing to output execute in close? Is the close operation heavy?

File: flink-table-store-common/src/main/java/org/apache/flink/table/store/lookup/hash/HashLookupStoreWriter.java [hunk 4]
```diff
@@ -0,0 +1,486 @@
...
+                                + slotSize
+                                + " bytes";
+
+                LOG.info("Built index file {}\n" + msg, indexFile.getName());
+            } finally {
+                // Close input
+                tempIndexStream.close();
+
+                // Close index and make sure resources are liberated
+                indexChannel.close();
```
Comment (SteNicholas) [inline flink-table-store-common/src/main/java/org/apache/flink/table/store/lookup/hash/HashLookupStoreWriter.java:349]: Does the `byteBuffer` need to clean?

File: flink-table-store-common/src/main/java/org/apache/flink/table/store/lookup/hash/HashLookupStoreWriter.java [hunk 5]
```diff
@@ -0,0 +1,486 @@
...
+            dos = new DataOutputStream(new BufferedOutputStream(new FileOutputStream(file)));
+            indexStreams[keyLength] = dos;
+
+            dataLengths[keyLength]++;
+        }
+        return dos;
+    }
+
+    private int getNumKeyCount() {
+        int res = 0;
```
Comment (SteNicholas) [inline flink-table-store-common/src/main/java/org/apache/flink/table/store/lookup/hash/HashLookupStoreWriter.java:478]:
```
return Arrays.stream(keyCounts).filter(c -> c != 0).count;
```

File: flink-table-store-common/src/main/java/org/apache/flink/table/store/utils/SimpleReadBuffer.java [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+
+package org.apache.flink.table.store.utils;
+
+/** A simple read buffer provide {@code readUnsignedByte} and position. */
```
Comment (SteNicholas) [inline flink-table-store-common/src/main/java/org/apache/flink/table/store/utils/SimpleReadBuffer.java:17]:
```suggestion
/** A simple read buffer provides {@code readUnsignedByte} and position. */
```

File: flink-table-store-common/src/main/java/org/apache/flink/table/store/lookup/hash/HashLookupStoreWriter.java (hunk 3 above)
Comment (JingsongLi) [reply flink-table-store-common/src/main/java/org/apache/flink/table/store/lookup/hash/HashLookupStoreWriter.java:178]: Closing just copy files into one file. It is not so heavy.

File: flink-table-store-common/src/main/java/org/apache/flink/table/store/lookup/hash/HashLookupStoreWriter.java (hunk 4 above)
Comment (JingsongLi) [reply flink-table-store-common/src/main/java/org/apache/flink/table/store/lookup/hash/HashLookupStoreWriter.java:349]: `FileChannel` is closed, there is no close in `byteBuffer`.

File: flink-table-store-common/src/main/java/org/apache/flink/table/store/lookup/hash/HashLookupStoreWriter.java (hunk 5 above)
Comment (JingsongLi) [reply flink-table-store-common/src/main/java/org/apache/flink/table/store/lookup/hash/HashLookupStoreWriter.java:478]: In performance critical path, we don't use java stream.

**Review by SteNicholas — APPROVED** [review]
LGTM.

**Pull request merged**